### PR TITLE
EasyAuth Support

### PR DESCRIPTION
**Pull request by mathewc**
Addresses https://github.com/Azure/azure-webjobs-sdk-script/issues/33 and https://github.com/Azure/azure-webjobs-sdk-script/issues/926. PR for the supporting binding changes can be found here: https://github.com/Azure/azure-webjobs-sdk-extensions/pull/262.

With these changes, we allow users to specify a new "user" auth level value to indicate that a function requires an EasyAuth validated identity.

We also now expose KeyId information to user code via the same mechanism.

I'm still in the process of adding some more tests.

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 1]
```diff
@@ -59,17 +64,58 @@ public FunctionsController(WebScriptHostManager scriptHostManager, WebHookReceiv
...
             var authorizationLevel = authorizationResult.AuthorizationLevel;
+
+            if (authorizationLevel > AuthorizationLevel.Anonymous)
+            {
+                // add a new identity to the principal representing
+                // the key authentication result
+                AddKeyAuthorizationIdentity(ClaimsPrincipal.Current, authorizationResult);
+            }
+
+            // See if we're we're authorized at the User level (EasyAuth)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: We want to determine a single auth level for the request but must deal with the possibility that a request might include both a key as well as EasyAuth credentials.

File: src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj [hunk 2]
```diff
@@ -189,7 +189,7 @@
       <HintPath>..\..\packages\Microsoft.Azure.WebJobs.Core.2.1.0-beta1-10950\lib\net45\Microsoft.Azure.WebJobs.dll</HintPath>
     </Reference>
     <Reference Include="Microsoft.Azure.WebJobs.Extensions, Version=2.1.0.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Microsoft.Azure.WebJobs.Extensions.2.1.0-beta1-10467\lib\net45\Microsoft.Azure.WebJobs.Extensions.dll</HintPath>
+      <HintPath>C:\code\Git\Azure\azure-webjobs-sdk-extensions\test\WebJobs.Extensions.Tests\bin\Debug\Microsoft.Azure.WebJobs.Extensions.dll</HintPath>
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/WebJobs.Script.WebHost.csproj]: Temporary changes for local e2e testing that will be reverted

File: src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs [hunk 3]
```diff
@@ -465,6 +469,29 @@ protected override void OnScriptFileChanged(FileSystemEventArgs e)
...
+            {
+                { "authenticationType", identity.AuthenticationType }
+            };
+
+            var claims = new Dictionary<string, object>();
+            foreach (var currClaim in identity.Claims)
+            {
+                var claim = new Dictionary<string, object>
+                {
+                    { "type", currClaim.Type },
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs]: Need to finish this and add unit tests

File: src/WebJobs.Script/Description/ScriptFunctionInvokerBase.cs [hunk 4]
```diff
@@ -170,6 +171,31 @@ protected virtual void InitializeEnvironmentVariables(Dictionary<string, string>
...
+            InitializeIdentityEnvironmentVariables(environmentVariables, context.Identity);
+        }
+
+        internal static void InitializeIdentityEnvironmentVariables(Dictionary<string, string> environmentVariables, ClaimsIdentity identity)
+        {
+            environmentVariables["IDENTITY_AUTHENTICATION_TYPE"] = identity.AuthenticationType;
+
+            foreach (var claim in identity.Claims)
+            {
+                // some script environments have problems with special characters in
```
Comment (mathewc) [inline src/WebJobs.Script/Description/ScriptFunctionInvokerBase.cs]: Anyone have any suggestions on alternate ways of handling this? I ran into this issue in my testing - claim names can be very long and contain characters that are special in languages like Batch.

File: src/WebJobs.Script/ScriptConstants.cs [hunk 5]
```diff
@@ -22,6 +22,10 @@ public static class ScriptConstants
         public const string TracePropertyIsUserTraceKey = "MS_IsUserTrace";
         public const string TracePropertyIsSystemTraceKey = "MS_IsSystemTrace";
 
+        public const string AzureFunctionsKeyAuthenticationType = "key";
+        public const string AzureFunctionsAuthLevelClaimName = "urn:functions:authLevel";
```
Comment (mathewc) [inline src/WebJobs.Script/ScriptConstants.cs]: Open to suggestions on the naming scheme for these claims

File: src/WebJobs.Script/Description/ScriptFunctionInvokerBase.cs (hunk 4 above)
Comment (cgillum) [inline src/WebJobs.Script/Description/ScriptFunctionInvokerBase.cs]: Don't we often write to a file for script-based languages like Batch?

File: sample/WebHook-Generic-CSharp/run.csx [hunk 6]
```diff
@@ -2,9 +2,14 @@
 
 using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
+using System.Security.Claims;
 
-public static JObject Run(Payload payload, string action)
+public static JObject Run(Payload payload, string action, ClaimsIdentity identity, TraceWriter log)
```
Comment (mattchenderson) [inline sample/WebHook-Generic-CSharp/run.csx]: Another instance where I wonder if we should leave the basic sample alone, and have a separate auth sample.
Also, webhooks generally don't include a user identity, so I'm not sure how valid this scenario is.

File: src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs [hunk 7]
```diff
@@ -152,7 +150,7 @@ internal static ScriptHostConfiguration CreateScriptHostConfiguration(WebHostSet
 
         private static void InitializeFileSystem(string scriptPath)
         {
-            if (ScriptSettingsManager.Instance.IsAzureEnvironment)
+            if (ScriptSettingsManager.Instance.IsAppServiceEnvironment)
```
Comment (mattchenderson) [inline src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs]: Nit: AppServiceEnvironment has a specific meaning which I don't think is intended here.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs (hunk 1 above)
Comment (mattchenderson) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: I'm still not sure that I like that assertion. If the developer specifies "user" then that's clearly the most important thing to them. To ignore that in favor of a key seems like a violation of expectation. The user token is also a stronger security claim, so I think we want to enforce it when we're asked to. The point at which a shared secret is going into a client that a user interacts with is generally the point at which it should no longer be considered a secret.

I think we'd also discussed splitting keys and user config settings to allow composite authorization checks. That would mean that "anonymous" would have to be renamed (arguably, it should be anyway).

**Comment by mathewc** [comment]
Update on this PR - It's on pause until the .NET Core migration is complete, at which time I'll rebase this and address all the feedback given thus far.

**Comment by mathewc** [comment]
Thanks @mattchenderson for the feedback. I'm closing this PR since it needs to be ported to .NET Core so will undergo significant changes. However, when the work resumes, I'll be sure to apply your feedback.

**Pull request closed without merging**